### DISCUSSION TITLE
Modals and buttons use the entity ID from the URL, not from the parameters passed.

### DISCUSSION THREAD

**Discussion by stas12go** [orchidsoftware/platform#2557] [Q&A]
I have URL like http://localhost/admin/product/137/reviews.
In this screen I operate with Review entities, not Product entity.
So when I try to (CR)UD Review entity (with ID 465 for example) I get Product ID (137 from URL) in my methods.

Screen looks like this.
![2023-03-15 12 59 13 localhost f9d34c176c5b](https://user-images.githubusercontent.com/78864187/225244336-0abbee55-ff02-4b35-9f57-914d36a5cc92.png)

For control buttons I have code:
```php
TD::make('controls', 'Controls')
    ->width('100')
    ->render(function (TelegramUserReview $model) {
        $fakeReviewButtonGroup = [
            ModalToggle::make('Edit')
                ->modal('EditReviewModal')
                ->modalTitle('Edit fake review')
                ->method('saveReview')
                ->asyncParameters(['id' => $model->id])
                ->icon('note'),
            Button::make('Delete')
                ->confirm('Do you want to delete the fake review?')
                ->method('deleteReview')
                ->parameters(['id' => $model->id])
                ->icon('trash'),
        ];

        $realReviewButtonGroup = [
            Button::make('Viewed')
                ->confirm('Do you want to mark review as Viewed?')
                ->method('markReviewAsViewed')
                ->parameters(['id' => $model->id])
                ->icon('check'),
        ];

        return Group::make($model->telegram_user_id ? $realReviewButtonGroup : $fakeReviewButtonGroup);
    })
```
As you can see I pass TelegramUserReview ID as parameter for `deleteReview` method. But when I dump it I get 137 instead of 465.
```php
/**
 * @param int $id
 * @return void
 */
public function deleteReview(int $id): void
{
    dd($id);
    TelegramUserReview::destroy($id);
}
```
![image](https://user-images.githubusercontent.com/78864187/225245592-eba0c5c0-7f5f-46be-9543-7e79df05c756.png)
I am pretty sure that review ID is 465, not 137.
![image](https://user-images.githubusercontent.com/78864187/225246005-561fa63d-038f-47e8-b42d-42d337526b8f.png)

Also I have URL like http://localhost/admin/reviews. This URL uses the same screen.
Deleting the review works properly here. I mean I get correct review ID in `deleteReview` method.

Same thing happens with Modals. I solved it by making hidden Input with review id, but it's too ugly solution.
```php
Layout::modal('EditReviewModal', [
    Layout::rows([
        TextArea::make('review')
            ->rows(5)
            ->title(__('Review'))
            ->placeholder(__('Review')),
        Input::make('id')
            ->hidden(),
    ]),
])
    ->async('asyncGetReview')
    ->applyButton(__('Save'))
...
/**
 * @param Request $request
 * @return void
 */
public function saveReview(Request $request): void
{
    $this->validateReviewRequest($request);

    $review = TelegramUserReview::query()->findOrFail((int)$request->input('id'));

    $review->review = $request->input('review');

    if ($review->save()) {
        Toast::success('Fake review was successfully saved!');
    } else {
        Toast::error('Fake review was not saved! Try again later!');
    }
}

```

Have you guys seen this in your code? How did you solve that?

**Comment by c-gross** [comment]
Depends whats your model at this point when you render the content and buttons. Since the query or the table content is not in the code snippets, it's difficult to say if model is the review or product object.

Maybe it's just something like: $model->review->id

**Reply by stas12go** [reply]
Hi!
Here is [Gist](https://gist.github.com/stas12go/8d2db833683427dcab742a63b2c3ff98).

Actually I have one product model and two review models.

I use review model in closure in `render()` method for Controls TD, so I have to use `$model->id`.

**Reply by c-gross** [reply]
Yes, looks correct. I had something similar but was working for me.

'm loading an order and all orderitem in the query. Had a modal to update something for orderitems and a link to a different route using the orderitem id. The correct id is passed in my case:

```
Link::make(__('Print Label'))
    ->icon('fa.print-solid')
    ->canSee(Auth::user()->hasAccess('platform.production'))
    ->target('_blank')
    ->route('platform.my.route.name', $orderItem->id),
ModalToggle::make('Edit Customization')
    ->modalTitle('Edit Customization')
    ->modal('editCustomization')
    ->method('saveCustomization')
    ->asyncParameters([
        'orderItem' => $orderItem->id
    ])
    ->icon('pencil'),
```

**Reply by stas12go** [reply]
Can you please share whole class?

**Reply by c-gross** [reply]
I remove the order table and some other stuff since it was a but file:
https://gist.github.com/c-gross/14d0902dfefe610d553df1cdd6d9d8f4

**Comment by stas12go** [comment]
I noticed that `\Orchid\Screen\Resolvers\ScreenDependencyResolver::resolveScreen` method prepare parameters for my `deleteReview` method:
```php
/**
 * @param \Orchid\Screen\Screen $screen
 * @param string                $method
 * @param array                 $httpQueryArguments
 *
 * @throws \ReflectionException
 *
 * @return array
 */
public function resolveScreen(Screen $screen, string $method, array $httpQueryArguments = []): array
{
    $parameters = (new ReflectionClass($screen))->getMethod($method)->getParameters();

    $arguments = collect($httpQueryArguments);

    return collect($parameters)
        ->map(function (ReflectionParameter $parameter) use (&$arguments) {
            return $this->bind($parameter, $arguments);
        })
        ->all();
}
```
In `$parameters` variable I get array of `deleteReview` parameters 
```php
[
    0 => $id
]
```
and in `$arguments` I get array like 
```php
[
    0 => {productId},
    1 => 'deleteReview',
    'id' => {reviewId that I pass in Button `parameters()` method}
]
```
So If I add two more parameters to `deleteReview` signature (in order like in `$arguments` array above), it works properly:
```php
/**
 * @param int $productId
 * @param string $method
 * @param int $reviewId
 * @return void
 */
public function deleteReview(int $productId, string $method, int $reviewId): void
{
    TelegramUserReview::destroy($reviewId);
}
```
But I dont think it's clear code.
Probably I have to redefine `\Orchid\Screen\Screen::handle` method in my `app/Orchid/Screens/TelegramUsers/TelegramUsersReviewsScreen.php` file and make a child class for `\Orchid\Screen\Resolvers\ScreenDependencyResolver` and redefine `resolveScreen()` method there.

Do you think the latter approach is a good idea?

P.s. I just tested this. It works. All I had to do was add `->reverse()` to the expression `$arguments = collect($httpQueryArguments)
;`.

**Comment by zyz954489346** [comment]
mark
